### PR TITLE
Feature: Added logistic regression

### PR DESCRIPTION
**Pull request by KennethEnevoldsen**
- Added logistic regression
- Added imputation parameter for models 
- Updated synthetic datasset to include NAs
- Added a simple Imputer for imputing NAs

### CODE REVIEW

File: src/psycopt2d/synth_data_generator/synth_data_generator.py [hunk 1]
```diff
@@ -1,52 +1,102 @@
...
+    noise_mean_sd: Tuple[float, float] = (0, 1),
+) -> pd.DataFrame:
     """Takes a dict and generates synth data from it.
 
     Args:
-        dict (Dict): A dict representing each column of shape: {"col_name": {"column_type": str, "output_type": ["float"|"int"], "min": int, "max": int}}
-        len (int): Number of rows to generate.
+        predictors (Dict): A dict representing each column of shape: {"col_name":
+            {"column_type": str, "output_type": ["float"|"int"], "min": int, "max":
+            int}}
```
Comment (MartinBernstorff) [inline src/psycopt2d/synth_data_generator/synth_data_generator.py:24]:
```suggestion
{"col_name":
            {"column_type": str, 
            "output_type": ["float"|"int"], 
            "min": int, 
            "max": int
            }
        }
```

File: src/psycopt2d/synth_data_generator/synth_data_generator.py [hunk 2]
```diff
@@ -1,52 +1,102 @@
...
+        n_samples (int): Number of samples (rows) to generate.
+        synthetic_model (str): the syntehtic model e.g. specified as
+            '1*col_name+1*col_name2'
```
Comment (MartinBernstorff) [inline src/psycopt2d/synth_data_generator/synth_data_generator.py:27]:
```suggestion
        logistic_outcome_model (str): The statistical model used to generate outcome values, e.g. specified as'1*col_name+1*col_name2'
```

Comment (MartinBernstorff) [reply src/psycopt2d/synth_data_generator/synth_data_generator.py:27]: Not sure this is the best name for it, but we probably want something in this vein.

File: src/psycopt2d/synth_data_generator/synth_data_generator.py [hunk 3]
```diff
@@ -1,52 +1,102 @@
...
+        na_prob (float, optional): probability of changing a predictor column to NA
```
Comment (MartinBernstorff) [inline src/psycopt2d/synth_data_generator/synth_data_generator.py:28]:
```suggestion
        na_prob (float, optional): Probability of changing a value in a predictor column to NA.
```

File: src/psycopt2d/synth_data_generator/synth_data_generator.py [hunk 4]
```diff
@@ -1,52 +1,102 @@
...
+        prob_outcome (float): probability of outcome
```
Comment (MartinBernstorff) [inline src/psycopt2d/synth_data_generator/synth_data_generator.py:29]:
```suggestion
        prob_outcome (float): Probability of a given row receiving "1" for the outcome.
```

File: src/psycopt2d/synth_data_generator/synth_data_generator.py [hunk 5]
```diff
@@ -1,52 +1,102 @@
...
+        noise_mean_sd (Tuple[float, float], optional): mean and sd of the noise.
+            Increase SD to optain more uncertain models.
```
Comment (MartinBernstorff) [inline src/psycopt2d/synth_data_generator/synth_data_generator.py:31]:
```suggestion
            Increase SD to obtain more uncertain models.
```

File: src/psycopt2d/train_model.py [hunk 6]
```diff
@@ -126,7 +133,6 @@ def stratified_cross_validation(
 @hydra.main(
     config_path=CONFIG_PATH,
     config_name="default_config",
-    version_base="1.2",
```
Comment (MartinBernstorff) [inline src/psycopt2d/train_model.py:129]:
```suggestion
    version_base="1.2",
```

Deleting this causes warnings, we probably want to have a minimum version of hydra. Any reason for deleting? :-)

File: src/psycopt2d/train_model.py [hunk 7]
```diff
@@ -200,11 +206,6 @@ def main(cfg):
 
     run.finish()
 
-    return roc_auc_score(
```
Comment (MartinBernstorff) [inline src/psycopt2d/train_model.py:203]:
```suggestion
    return roc_auc_score(
        eval_dataset[OUTCOME_COL_NAME],
        eval_dataset[y_hat_prob_col_name],
    )

```

We need to return a metric for Optuna :-)

**Review by MartinBernstorff — COMMENTED** [review]
Looks very good, Kenneth! Super duper modular, and nice extra functionality in `synth_data_generator`!

A few comments to `train_model` that are important.

File: src/psycopt2d/train_model.py (hunk 6 above)
Comment (KennethEnevoldsen) [reply src/psycopt2d/train_model.py:129]: It is not deleted

File: src/psycopt2d/config/model/ebm.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+# Parameters that will only take effect if running with --multirun
+hydra:
+  sweeper:
+    params:
+      ++model.args.interactions: choice(0, 5, 10, 50)
+      ++model.args.learning_rate: interval(0.001, 0.1)
+      ++model.args.validation_size: interval(0.20, 0.05)
+      ++model.args.min_samples_leaf: choice(1, 2)
+      ++model.args.max_leaves: choice(2, 3, 4, 8, 16)
+      ++model.args.l1_ratio: interval(1e-8, 1.0)
```
Comment (MartinBernstorff) [inline src/psycopt2d/config/model/ebm.yaml:32]: l1_ratio isn't specified in the base model above :-)

File: src/psycopt2d/config/model/naive-bayes.yaml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# @package _global_
+model:
+  model_name: naive-bayes # (str): Model name
+  require_imputation: True # (bool): Whether the model requires imputation.
+  args:   # https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html
```
Comment (MartinBernstorff) [inline src/psycopt2d/config/model/naive-bayes.yaml:5]:
```suggestion
  args:   # https://scikit-learn.org/stable/modules/naive_bayes.html#gaussian-naive-bayes
```

File: src/psycopt2d/config/preprocessing/default_preprocessing.yaml [hunk 10]
```diff
@@ -1,2 +1,4 @@
 convert_to_boolean: False # (Boolean): Convert all prediction values (except gender) to boolean. Defaults to False
-convert_datetimes_to: False # (str): Options include ordinal or False 
\ No newline at end of file
+convert_datetimes_to: False # (str): Options include ordinal or False 
+imputation_method: "most_frequent" # (str): Options include 2most_frequent"
+transform: null # (str|null): transformation applied to all predictors after imputation. options include "z-score-normalization"
```
Comment (MartinBernstorff) [inline src/psycopt2d/config/preprocessing/default_preprocessing.yaml:4]:
```suggestion
transform: null # (str|null): Transformation applied to all predictors after imputation. Options include "z-score-normalization"
```

File: src/psycopt2d/synth_data_generator/synth_data_generator.py [hunk 11]
```diff
@@ -1,52 +1,115 @@
...
+                "min": int,
+                "max":int
+                }
+            }
+        n_samples (int): Number of samples (rows) to generate.
+        logistic_outcome_model (str): The statistical model used to generate outcome
+            values, e.g. specified as'1*col_name+1*col_name2'
+        na_prob (float, optional): Probability of changing a value in a predictor column
+            to NA.
+        na_ignore_cols (List[str], optional): columns to ignore when creating NAs
```
Comment (MartinBernstorff) [inline src/psycopt2d/synth_data_generator/synth_data_generator.py:37]:
```suggestion
        na_ignore_cols (List[str], optional): Columns to ignore when creating NAs
```

File: tests/test_train_model.py [hunk 12]
```diff
@@ -3,8 +3,28 @@
...
+
+
+def test_main_ebm():
+    """test main using an explainable boosting machine."""
+    with initialize(version_base=None, config_path="../src/psycopt2d/config/"):
+        cfg = compose(
+            config_name="test_config.yaml",
+            overrides=["+model=ebm"],
+        )
+        main(cfg)
```
Comment (MartinBernstorff) [inline tests/test_train_model.py:30]:
```suggestion
def test_main_naive_bayes():
    """test main using an explainable boosting machine."""
    with initialize(version_base=None, config_path="../src/psycopt2d/config/"):
        cfg = compose(
            config_name="test_config.yaml",
            overrides=["+model=naive-bayes"],
        )
        main(cfg)

```

**Review by MartinBernstorff — COMMENTED** [review]
Tiny modifications, otherwise LGTM!
